web/models/product.py

```diff
@@ -122,16 +122,37 @@ def get_metadata(self, description, graph, terms=None):
         for term in terms:
 
             # Generate unstemmed ngrams of the same length as the product match
+            remaining_tokens = []
             n = len(term)
+            tag = 0
             for tokens in tokenize(
                 doc=description,
                 ngrams=n,
                 stemmer=None,
                 analyzer=self.analyzer
             ):
+                # If generated tokens are depleted, consume remaining tokens
+                if len(tokens) < n and len(remaining_tokens) > 0:
+                    tokens = remaining_tokens
+
+                # Continue token-by-token advancement, closing any open tags
+                tag -= 1
+                if tag == 0:
+                    markup += '</mark>'
```

Comment:
As an observation: it looks like most if not all of these operations could be handled by a client reading these events as a stream; it may be possible to extract a generator-based `get_markup` method.